### ISSUE TITLE
Question regarding default soundpack

### ISSUE THREAD

**Issue by nmeum** [mumble-voip/mumble#4492]
At Alpine Linux downstream it was recently reported to us that mumble 1.3.2 emits error such as `<W>2020-09-18 14:26:33.482 File  ":/ServerConnected.ogg"  failed to open` when connecting to a server et cetera:  https://gitlab.alpinelinux.org/alpine/aports/-/issues/11933 and consequently does not emitted any sounds.

A similar bug report also exists in the Gentoo bug tracker: https://bugs.gentoo.org/show_bug.cgi?id=477068

I investigated this further and can confirm this issue. However, from the resources provided by the mumble project I do not fully understand how the default soundpack is supposed to work. Seems to me the sounds are intended as a fallback when TTS is not available. The default soundpack itself was added in 01a72bbf9f5a76153e54b741eb4e56d7e1599a00. However, it seems to me that these sounds are not installed by default by the mumble build system. Are they supposed to be installed manually?

Furthermore, to which path should these files be installed? Filepaths are obtained as `QLatin1String(":/wb_male.oga")` et cetera in the source code, which seems to refer to a relative filepath defined in `mumble.qrc`. However, I am not familiar with Qt and do not understand to which path they are relative to.

https://github.com/mumble-voip/mumble/blob/01a72bbf9f5a76153e54b741eb4e56d7e1599a00/src/mumble/mumble.qrc#L41-L49

I hope you don't mind me reaching out here but answers regarding the raised questions would be appreciated in order to allow me to fix the downstream bug.

**Comment by Krzmbrzl** [comment]
The paths are relative to the `.qrc` file they are referenced. Therefore `../../samples` refers to https://github.com/mumble-voip/mumble/tree/master/samples

To my understanding of [Qt's resource system](https://doc.qt.io/qt-5/resources.html) these sounds should get baked into the final application though, so they should only have to be found during compilation:
> The Qt resource system is a platform-independent mechanism for storing binary files in the application's executable. 

I don't quite understand how these sounds are not found given that they *should* be bundled in the executable :thinking:

**Comment by nmeum** [comment]
Here is our build log: https://build.alpinelinux.org/buildlogs/build-edge-x86_64/community/mumble/mumble-1.3.2-r1.log

Our [config](https://gitlab.alpinelinux.org/alpine/aports/-/blob/master/community/mumble/APKBUILD#L30):

```
CONFIG+="no-speechd no-g15 no-ice bundled-celt no-embed-qt-translations no-bundled-speex no-bundled-opus no-rnnoise"
```

not sure if that helps. Any idea what I could be looking for to debug this?

**Comment by Krzmbrzl** [comment]
I guess the first thing that I'd try to do would be to try and compile Mumble with default options (and if needed `no-g15` and `no-ice` should be safe as well) and see if the problem persists.
If it does not, then there is some of the set options have side-effects affecting the sounds. I especially suspect that `no-embed-qt-translations` might be such a candidate...

**Comment by nmeum** [comment]
Tried `CONFIG+="no-ice no-g15 no-speechd"` but the problem still seems to be reproducible, will try to reduce the config further. Looking at the buildlog the sound files are also correctly embedded in the binary:

```
/usr/lib/qt5/bin/rcc -compress 9 -name mumble mumble.qrc -o release/qrc_mumble.cpp
```

And `qrc_mumble.cpp` is later compiled to `qrc_mumble.o` and included in the linker command-line for the `mumble` binary. Primitive checks with `strings(1)` seem to confirm that the samples are included:

```
$ strings ./samples/off.ogg | grep libVorbis
Xiph.Org libVorbis I 20070622
$ strings `which mumble` | grep 'Xiph.Org libVorbis I 20070622'
Xiph.Org libVorbis I 20070622
Xiph.Org libVorbis I 20070622
Xiph.Org libVorbis I 20070622
```

Maybe this is just a problem with the lookup performed by `QLatin1String` (again: I don't know Qt)?

**Comment by Krzmbrzl** [comment]
> Maybe this is just a problem with the lookup performed by QLatin1String (again: I don't know Qt)?

If that was the case I'd expect this to be an issue on all systems, but for instance on my system (KDE Neon) it seems to work fine :eyes: 

Does this also happen when building mumble and running the built binary without the install step?
It might be also worth checking whether this issue still exists when building from current master instead of the 1.3.x branch (for which you'll have to use cmake though) :thinking:

**Comment by nmeum** [comment]
I debugged this further with gdb and managed to fix it :tada: This was strictly speaking not a mumble issue, but instead a problem with Alpine's current libsndfile package. See [!12659](https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/12659) for more information. I managed to discover this by debugging the `SoundFile` constructor with gdb. The `qfFile.open` invocation itself actually works fine, what fails is `sf_open_virtual`:

https://github.com/mumble-voip/mumble/blob/af7dac72f4063dd8d4dac71973ea51c25896089a/src/mumble/AudioOutputSample.cpp#L29-L34

I would highly recommend adding an explicit error check for that `sf_open_virtual` invocation. That would have made it much easier for me to discover the problem. I ultimately ended up with the following printf-debugging code:

```patch
diff -upr mumble-1.3.2.orig/src/mumble/AudioOutputSample.cpp mumble-1.3.2/src/mumble/AudioOutputSample.cpp
--- mumble-1.3.2.orig/src/mumble/AudioOutputSample.cpp  2020-07-09 19:06:18.000000000 +0200
+++ mumble-1.3.2/src/mumble/AudioOutputSample.cpp       2020-09-20 17:53:03.203494963 +0200
@@ -7,6 +7,8 @@
 
 #include "AudioOutputSample.h"
 
+#include <iostream>
+
 #include "Audio.h"
 
 SoundFile::SoundFile(const QString &fname) {
@@ -25,6 +27,8 @@ SoundFile::SoundFile(const QString &fnam
                static SF_VIRTUAL_IO svi = {&SoundFile::vio_get_filelen, &SoundFile::vio_seek, &SoundFile::vio_read, &SoundFile::vio_write, &SoundFile::vio_tell};
 
                sfFile = sf_open_virtual(&svi, SFM_READ, &siInfo, this);
+               if (!sfFile)
+                       std::cout << "sf_open_virtual failed: " << sf_strerror(NULL) << std::endl;
        }
 }
```

This bug can be closed though, I will just leave it open in case you want to implement a proper error check for `sf_open_virtual`.

Thanks a lot for your input, wouldn't have discovered this without it! :)

**Comment by Krzmbrzl** [comment]
You're welcome :) 
And thanks for sharing the idea of how debugging this kind of error can be made easier in the future. I have implemented a check that produces a warning in #4497

**Comment by nmeum** [comment]
Nice, thanks a lot!
